stan/math/prim/fun/grad_2F1.hpp

```diff
@@ -8,10 +8,60 @@
 #include <stan/math/prim/fun/fabs.hpp>
 #include <stan/math/prim/fun/inv.hpp>
 #include <stan/math/prim/fun/log.hpp>
+#include <stan/math/prim/fun/sign.hpp>
 #include <cmath>
 
 namespace stan {
 namespace math {
+namespace internal {
+/**
+ * Returns the intermediate calculations for the gradient of the
+ * hypergeometric function (2F1).
+ *
+ * @tparam T scalar type or `var_value`
+ * @tparam T_StdVec std::array of doubles or `var_values`
+ * @tparam T_StdVecInt array of int
+ * @tparam T_int integer
+ * @param[out] log_g_old log of previous gradient values
+ * @param[out] log_g_old_sign sign of previous gradient values
+ * @param[out] log_t_new log of power series calculation
+ * @param[out] log_t_new_sign sign of power series calcuation
+ * @param[in] p intermediate power series value
+ * @param[in] log_z see hypergeometric function 2F1 definition
+ * @param[in] log_t_old log of previous power series calculation
+ * @param[in] log_t_old_sign log of previous power series calculation
+ * @param[in] k loop iteration
+ * @param[in] a1 see hypergeometric function 2F1 definition
+ * @param[in] a2 see hypergeometric function 2F1 definition
+ * @param[in] b1 see hypergeometric function 2F1 definition
+ */
+template <typename T, typename T_StdVec, typename T_StdVecInt, typename T_int>
+void calc_lambda(T_StdVec& log_g_old, T_StdVecInt& log_g_old_sign, T& log_t_new,
+                 T_int& log_t_new_sign, const T& p, const T& log_z,
+                 const T& log_t_old, const T_int& log_t_old_sign, const int k,
+                 const T& a1, const T& a2, const T& b1) {
+  using ret_t = return_type_t<T, T_int, T_StdVec, T_StdVecInt>;
+
+  log_t_new += log(fabs(p)) + log_z;
+  log_t_new_sign = p >= 0.0 ? log_t_new_sign : -log_t_new_sign;
+
+  ret_t term
+      = log_g_old_sign[0] * log_t_old_sign * exp(log_g_old[0] - log_t_old)
+        + inv(a1 + k);
+  log_g_old[0] = log_t_new + log(fabs(term));
+  log_g_old_sign[0] = term >= 0.0 ? log_t_new_sign : -log_t_new_sign;
+
+  term = log_g_old_sign[1] * log_t_old_sign * exp(log_g_old[1] - log_t_old)
+         + inv(a2 + k);
+  log_g_old[1] = log_t_new + log(fabs(term));
+  log_g_old_sign[1] = term >= 0.0 ? log_t_new_sign : -log_t_new_sign;
+
+  term = log_g_old_sign[2] * log_t_old_sign * exp(log_g_old[2] - log_t_old)
+         - inv(b1 + k);
+  log_g_old[2] = log_t_new + log(fabs(term));
+  log_g_old_sign[2] = term >= 0.0 ? log_t_new_sign : -log_t_new_sign;
```

Comment:
Given that the same calculations are being applied across the different arguments, we could probably vectorise this using something like:
```cpp
  const auto& log_g_old_vec = as_array_or_scalar(log_g_old);
  const auto& log_g_old_sign_vec = as_array_or_scalar(log_g_old_sign);
  Eigen::Array<T, Eigen::Dynamic, 1> hyper_args(3);
  hyper_args << a1, a2, b1;

  Eigen::Array<T, Eigen::Dynamic, 1> term
    = log_g_old_sign_vec * log_t_old_sign * exp(log_g_old_vec - log_t_old)
       + inv(hyper_args + k);

  log_g_old_vec = log_t_new + log(fabs(term));
  log_g_old_sign_vec = sign(value_of_rec(term)) * log_t_new_sign;

```